src/solvers/bicgstab_solver.hpp

```diff
@@ -0,0 +1,276 @@
+//========================================================================================
+// (C) (or copyright) 2023. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef SOLVERS_BICGSTAB_SOLVER_HPP_
+#define SOLVERS_BICGSTAB_SOLVER_HPP_
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "interface/mesh_data.hpp"
+#include "interface/meshblock_data.hpp"
+#include "interface/state_descriptor.hpp"
+#include "kokkos_abstraction.hpp"
+#include "solvers/mg_solver.hpp"
+#include "solvers/solver_utils.hpp"
+#include "tasks/task_id.hpp"
+#include "tasks/task_list.hpp"
+
+
+namespace parthenon {
+
+namespace solvers {
+
+struct BiCGSTABParams { 
+  int max_iters = 10; 
+  Real residual_tolerance = 1.e-12;  
+  Real restart_threshold = -1.0;
+  bool precondition = true;
+  bool flux_correct = true;
+}; 
+
+#define BICGVARIABLE(base, varname)                                                      \
+  struct varname : public parthenon::variable_names::base_t<false> {                     \
+    template <class... Ts>                                                               \
+    KOKKOS_INLINE_FUNCTION varname(Ts &&...args)                                         \
+        : parthenon::variable_names::base_t<false>(std::forward<Ts>(args)...) {}         \
+    static std::string name() { return base::name() + "." #varname; }                    \
+  }
+
+template <class x, class rhs, class equations> 
+class BiCGSTABSolver {
+  BICGVARIABLE(x, rhat0);
+  BICGVARIABLE(x, v);
+  BICGVARIABLE(x, h);
+  BICGVARIABLE(x, s);
+  BICGVARIABLE(x, t);
+  BICGVARIABLE(x, r);
+  BICGVARIABLE(x, p);
+  BICGVARIABLE(x, u);
+ 
+ public:
+  BiCGSTABSolver(StateDescriptor *pkg, BiCGSTABParams params_in, equations eq_in = equations()) 
+      : preconditioner(pkg, MGParams(), eq_in), params_(params_in), 
+        iter_counter(0), eqs_()  { 
+    using namespace refinement_ops;
+    auto mu = Metadata(
+      {Metadata::Cell, Metadata::Independent, Metadata::FillGhost, Metadata::WithFluxes, Metadata::GMGRestrict});
+    auto m_no_ghost = Metadata({Metadata::Cell, Metadata::Derived, Metadata::OneCopy});
+    pkg->AddField(u::name(), mu);
```

Comment:
There could be a name collision here if multiple packages try to add bicgstabsolvers. You could add `Metadata::Private` to the metadata flags. That would prefix all the variable names with the package name for the purposes of string-based name resolution.

I guess since they're templated on `x` maybe that's not an issue?

(I don't *think* the types will overlap, because the solver class is templated on x, rhs, and equations.)

Reply:
There is potential for a name collision, but only if you instantiate two solvers with the same type `x` since the `u::name()` should give something like `x::name() + ".u"`. It probably makes sense to make the fields private, but even then if you have to solvers that work on the same variable in a package there would be a collision. How to deal with these internal fields in cases where there is more than one solver needs some more thought, since we maybe don't want to allocate separate internal fields for each one (although that depends on the exact use case).